### DISCUSSION TITLE
Hosting a native window in ContentPage, View, FrameworkElement, etc.

### DISCUSSION THREAD

**Discussion by whodges** [dotnet/maui#15889] [General]
Is it possible to host a native window in a ContentPage, View, FrameworkElement, etc. in MAUI?  Essentially, I want to a have a window hooked up to a Vulkan swap chain, and I can only create a Vulkan surface from a native window.

I believe I could make a ViewHandler that creates a SwapChainPanel (a DirectX swap chain), and then I could copy the rendered Vulkan framebuffer to the DX one, but that'll only work on Windows, and it's not the prettiest solution.

**Comment by mattleibow** [comment]
I think there is a Vulkan something for all the platforms:
 - Android: this is built into Android/NDK: https://developer.android.com/ndk/guides/graphics
 - Windows: this is possible as you know 
 - iOS/Mac Catalyst: this is done with MoltenVK which is a layer on top of Metal: https://github.com/KhronosGroup/MoltenVK

The great thing about .NET Maui is that if you get some platform view working using native/C# code, you can make it work in .NET MAUI. So if you experiment with creating a UIView or FrameworkElement render Vulkan graphics, you can then make that view work in .NET Maui using the ViewHandler as you mention.

I did see there is some work with "Ashes" and "Dozen" to make Vulkan over DX (like MoltenVK) but I have not done any research on those: https://www.vulkan.org/porting

But TL;DR - if you make a view work in a plain old iOS/Android/WinUI app, you can make it work in .NET Maui. 

Hopefully this helps with some ideas...

Also, for getting Vulkan everywhere, this is the goal of the project and they have many communities: https://www.vulkan.org/#join-our-community I am sure you already reached out, but just in case. Hopefully there is some guru that has been running his entire company of X and it just works!

**Reply by whodges** [reply]
Thanks!  I appreciate the reply.  This question was meant to be less about Vulkan, and more about whether or not it's possible to actually have native windows (i.e., a window with an HWND on Windows, an ANativeWindow on Android, and a UIView on iOS) implemented in a PlatformView.

In ViewHandler, the TPlatformView must be a FrameworkElement, so that precludes the PlatformView from *being* a native window.  But I was wondering if a FrameworkElement could *contain* a native window.  I'm not optimistic at the moment, but I'm going to keep experimenting.

**Reply by mattleibow** [reply]
With regards to hosting an HWND in a framework element, I would maybe ask the WinUI team about that: https://github.com/microsoft/microsoft-ui-xaml

I will also try get an answer and report back...

**Comment by r2d2Proton** [comment]
I have been studying and working with Vulkan as of late. Now testing .NET MAUI. Where should I start folding in 3D graphics to a window in a MAUI app? Looking for a project template source file or class name.
